Remove TSource default constructor requirement for IncrementalLoadingCollection

hermitdave opened the pull request:
Remove TSource default constructor requirement for IncrementalLoadingCollection
Addressing #616

Updated docs and usage of the collection

hermitdave commented:
@deltakosh something for you

skendrot commented on Microsoft.Toolkit.Uwp/IncrementalLoadingCollection/IncrementalLoadingCollection.cs:
```diff
@@ -118,27 +118,6 @@ private set
-        /// </param>
-        /// <seealso cref="IIncrementalSource{TSource}"/>
-        public IncrementalLoadingCollection(int itemsPerPage = 20, Action onStartLoading = null, Action onEndLoading = null, Action<Exception> onError = null)
-            : this(new TSource(), itemsPerPage, onStartLoading, onEndLoading, onError)
```
Removing this method is a breaking change. Existing code will fail with the new version. It should still allow for the source to be generated for the user.

hermitdave replied:
yes this has been discussed.. if you can find a way to instantiate TSource without forcing default constructor by means of new() constraint, I am all ears

hermitdave replied:
right reflection... let me give that a try

hermitdave replied:
@skendrot change submitted.. no longer a breaking change

skendrot commented on Microsoft.Toolkit.Uwp/IncrementalLoadingCollection/IncrementalLoadingCollection.cs:
```diff
@@ -159,6 +160,18 @@ public IncrementalLoadingCollection(int itemsPerPage = 20, Action onStartLoading
         /// <seealso cref="IIncrementalSource{TSource}"/>
         public IncrementalLoadingCollection(TSource source, int itemsPerPage = 20, Action onStartLoading = null, Action onEndLoading = null, Action<Exception> onError = null)
```
This ctor shouldn't be changed. The ctor that does not include the TSource should be changed to no call this ctor and to try to create the TSource

hermitdave replied:
the original code used 2nd constructor primarly and I am continuing to use it. What I have changed is the param passed to 2nd constructor. This isn't a valid objection

skendrot commented on Microsoft.Toolkit.Uwp.SampleApp/SamplePages/Incremental Loading Collection/IncrementalLoadingCollectionPage.xaml.cs:
```diff
@@ -41,7 +41,8 @@ protected override void OnNavigatedTo(NavigationEventArgs e)
             base.OnNavigatedTo(e);
 
             // IncrementalLoadingCollection can be bound to a GridView or a ListView. In this case it is a ListView called PeopleListView.
-            var collection = new IncrementalLoadingCollection<PeopleSource, Person>();
```
No reason to change this

skendrot commented on Microsoft.Toolkit.Uwp.SampleApp/SamplePages/Microsoft Graph Service/MicrosoftGraphPage.xaml.cs:
```diff
@@ -117,7 +117,9 @@ public MicrosoftGraphPage()
                 top = Convert.ToInt32(txtTop);
             }
 
+            var source = new MicrosoftGraphSource();
```
No reason to change this

skendrot replied:
Putting to code in the second allows for null to be passed to the constructor. This should be an ArgumentNullException for the ctor. Allow for backward compatibility with the current ctor and enforce proper arguments for the new ctor

hermitdave commented:
@skendrot most changes have been reverted.. 
The 2nd constructor allowed nulls and there was no handling. In this instance we are trying to intantiate Source failing which we through error.

deltakosh commented on Microsoft.Toolkit.Uwp/IncrementalLoadingCollection/IncrementalLoadingCollection.cs:
```diff
@@ -159,6 +160,18 @@ public IncrementalLoadingCollection(int itemsPerPage = 20, Action onStartLoading
+
+                source = (TSource)constructor.Invoke(null);
+            }
+
```
woot :)

Did you test it on current sample app (Microsoft graph and CollectionSource sample )?

deltakosh replied:
If it was tested and it worked we can merge

hermitdave replied:
I have reverted all changes to Microsoft graph sample and the collection sample though I tested the collection sample with and without all changes

hermitdave replied:
@deltakosh I tested it so TSource had a default constructor and it was instantiated and used.. then I added a param to constructor and ensure that it failed correctly and I then commited the code

deltakosh replied:
Ok just need formal approval from @skendrot then

hermitdave commented:
@deltakosh I have modified and moved the code to instantiate instance of source using reflection outside of constructor. This is should placate any issues @skendrot had
